proxy/ParentConsistentHash.cc

```diff
@@ -64,23 +66,117 @@ ParentConsistentHash::getPathHash(HttpRequestData *hrdata, ATSHash64 *h)
   const char *tmp = NULL;
   int len;
   URL *url = hrdata->hdr->url_get();
+  char buffer[1024];
+  int slen     = 0;
+  int num_dirs = 0;
+
+  // Use over-ride URL from HttpTransact::State's cache_info.parent_selection_url, if present.
+  URL *ps_url = NULL;
+  Debug("parent_select", "ParentConsistentHash::%s() hrdata->cache_info_parent_selection_url = 0x%lx", __func__,
+        (unsigned long int)hrdata->cache_info_parent_selection_url);
+  if (hrdata->cache_info_parent_selection_url) {
+    ps_url = *(hrdata->cache_info_parent_selection_url);
+    Debug("parent_select", "ParentConsistentHash::%s() ps_url = 0x%lx", __func__, (unsigned long int)ps_url);
+    if (ps_url) {
+      tmp = ps_url->string_get_ref(&len);
+      if (tmp && len > 0) {
+        // Print the over-ride URL
+        slen = (len > 1023) ? 1023 : len;
+        strncpy(buffer, tmp, slen);
+        buffer[slen] = 0;
+        Debug("parent_select", "ParentConsistentHash::%s() Using Over-Ride String='%s'.", __func__, buffer);
+        h->update(tmp, len);
+        goto done;
+      }
+    }
+  }
 
   // Always hash on '/' because paths returned by ATS are always stripped of it
   h->update("/", 1);
 
   tmp = url->path_get(&len);
-  if (tmp) {
+
+  if (tmp && len > 0) {
+    // Print Original Path
```

Comment:
Perhaps include "ts/Diags.h" and use:

if (is_debug_tag_set("parent_select") {
  slen = (len > 1023) ? 1023 : len;
  strncpy(buffer, tmp, slen);
  buffer[slen] = 0;
  Debug("parent_select", "ParentConsistentHash::%s() Original Path='%s'.", __func__, buffer);
}